### PR TITLE
Build repositories content filtering

### PR DESCRIPTION
**Pull request by ljacomet**

### CODE REVIEW

**Comment by ljacomet** [comment]
@bot-gradle test QF

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

**Comment by ljacomet** [comment]
@bot-gradle test this

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc [hunk 1]
```diff
@@ -362,6 +362,14 @@ include::sample[dir="snippets/dependencyManagement/declaringRepositories-filteri
 It is possible to filter either by explicit _group_, _module_ or _version_, either strictly or using regular expressions.
 See link:{javadocPath}/org/gradle/api/artifacts/repositories/InclusiveRepositoryContentDescriptor.html[InclusiveRepositoryContentDescriptor] for details.
 
+[NOTE]
+====
+If you leverage exclusive content filtering in the <<plugins.adoc#sec:custom_plugin_repositories,`pluginManagement` section of the `settings.gradle(.kts)`>>, it becomes illegal to add more repositories through the project `buildscript.repositories`.
+In that case, the build configuration will fail.
+
+Your options are either to declare all repositories in settings or to use non-exclusive content filtering.
+====
```
Comment (JLLeitschuh) [inline subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc:371]: Link to the advisory?

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenRepositoryContentDescriptor.java [hunk 2]
```diff
@@ -65,4 +66,25 @@ public void snapshotsOnly() {
...
+            copy.setExcludedConfigurations(Sets.newHashSet(getExcludedConfigurations()));
+        }
+        if (getIncludeSpecs() != null) {
+            copy.setIncludeSpecs(Sets.newHashSet(getIncludeSpecs()));
+        }
+        if (getExcludeSpecs() != null) {
+            copy.setExcludeSpecs(Sets.newHashSet(getExcludeSpecs()));
+        }
+        if (getRequiredAttributes() != null) {
+            copy.setRequiredAttributes(Maps.newHashMap(getRequiredAttributes()));
```
Comment (JLLeitschuh) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenRepositoryContentDescriptor.java:86]: I think a deeper copy needs to occur here.

1. Is `Attribute` mutable?
2. The value in this map is itself a `Set<Object>`. Doesn't the value for every key need to be copied to a new `Set`?

**Review by JLLeitschuh — CHANGES_REQUESTED** [review]

File: subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java [hunk 3]
```diff
@@ -148,6 +150,10 @@ private void applyLegacyPlugins(@Nullable PluginManagerInternal target, Map<Resu
     }
 
     private void addPluginArtifactRepositories(RepositoryHandler repositories) {
+        if (pluginRepositoriesProvider.isExclusiveContentInUse() && !repositories.isEmpty()) {
+            throw new InvalidUserCodeException("When using exclusive repository content in 'settings.pluginManagement.repositories', you cannot add repositories to 'buildscript.repositories'.\n" +
+                "See the documentation in " + new DocumentationRegistry().getDocumentationFor("declaring_repositories", "declaring_content_exclusively_found_in_one_repository") + ".");
+        }
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/plugin-use/src/main/java/org/gradle/plugin/use/internal/DefaultPluginRequestApplicator.java:156]: I had a look but that was a too big adventure IMO

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenRepositoryContentDescriptor.java (hunk 2 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/DefaultMavenRepositoryContentDescriptor.java:86]: 1. No, it is not.
2. In theory, but in practice once that is realised, it is effectively immutable. If it was mutable, subsequent invocation of the content filter for a single repo could lead to inconsistent filtering. So by design it has to be effectively immutable.

File: subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc (hunk 1 above)
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/declaring_repositories.adoc:371]: I do not think it is relevant here.

**Comment by ljacomet** [comment]
@bot-gradle test this

**Review by JLLeitschuh — APPROVED** [review]
LGTM!

**Pull request merged**